File: src/main/java/hiiragi283/material/network/HiiragiNetworkManager.java
package hiiragi283.material.network;

import hiiragi283.material.RMReference;
import hiiragi283.material.tile.TileEntityModuleMachine;
import hiiragi283.material.util.OptionalUtil;
import net.minecraft.client.Minecraft;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;
import net.minecraftforge.fml.common.network.NetworkRegistry;
import net.minecraftforge.fml.common.network.simpleimpl.IMessage;
import net.minecraftforge.fml.common.network.simpleimpl.SimpleNetworkWrapper;
import net.minecraftforge.fml.relauncher.Side;
import org.jetbrains.annotations.Nullable;

import javax.annotation.ParametersAreNonnullByDefault;
import java.util.function.Consumer;

/**
 * Thanks to SkyTheory!
 * <a href="https://github.com/SkyTheory/SkyTheoryLib/blob/1.12.2/java/skytheory/lib/network/tile/TileSyncHandler.java">: Source</a>
 */

@ParametersAreNonnullByDefault
public abstract class HiiragiNetworkManager {

    public static SimpleNetworkWrapper INSTANCE = NetworkRegistry.INSTANCE.newSimpleChannel(RMReference.MOD_ID);

    public static void register() {
        INSTANCE.registerMessage(
                (message, ctx) -> syncMessage(Minecraft.getMinecraft().world, message.pos, message),
                HiiragiMessage.Client.class,
                0,
                Side.CLIENT
        );
        INSTANCE.registerMessage(
                (message, ctx) -> syncMessage(ctx.getServerHandler().player.world, message.pos, message),
                HiiragiMessage.Server.class,
                1,
                Side.SERVER
        );
        INSTANCE.registerMessage(
                (message, ctx) -> {
                    OptionalUtil.getTile(Minecraft.getMinecraft().world, message.pos, TileEntityModuleMachine.class)
                            .ifPresent(tile -> {
                                switch (message.tankIndex) {
                                    case 1 -> tile.tankInput1.setFluid(message.getFluidStack().orElse(null));
                                    case 2 -> tile.tankInput2.setFluid(message.getFluidStack().orElse(null));
                                    case 3 -> tile.tankOutput0.setFluid(message.getFluidStack().orElse(null));
                                    case 4 -> tile.tankOutput1.setFluid(message.getFluidStack().orElse(null));
                                    case 5 -> tile.tankOutput2.setFluid(message.getFluidStack().orElse(null));
                                    default -> tile.tankInput0.setFluid(message.getFluidStack().orElse(null));
                                }
                            });
                    return null;
                },
                HiiragiFluidMessage.class,
                2,
                Side.CLIENT
        );
    }

    @Nullable
    private static IMessage syncMessage(World world, BlockPos pos, HiiragiMessage message) {
        return syncMessage(world, pos, message, tile -> tile.readFromNBT(message.tag));
    }

    @Nullable
    private static IMessage syncMessage(World world, BlockPos pos, HiiragiMessage message, Consumer<TileEntity> consumer) {
        if (world.isBlockLoaded(pos)) OptionalUtil.getTile(world, pos).ifPresent(consumer);
        return null;
    }

}

File: src/main/kotlin/hiiragi283/material/network/HiiragiNetworkManager.kt
package hiiragi283.material.network

import hiiragi283.material.RMReference
import hiiragi283.material.entity.EntityMinecartTank
import hiiragi283.material.tile.TileEntityModuleMachine
import hiiragi283.material.util.getTile
import net.minecraft.client.Minecraft
import net.minecraft.entity.Entity
import net.minecraft.tileentity.TileEntity
import net.minecraft.world.World
import net.minecraftforge.fluids.FluidStack
import net.minecraftforge.fml.common.network.NetworkRegistry
import net.minecraftforge.fml.common.network.simpleimpl.IMessage
import net.minecraftforge.fml.common.network.simpleimpl.MessageContext
import net.minecraftforge.fml.common.network.simpleimpl.SimpleNetworkWrapper
import net.minecraftforge.fml.relauncher.Side

val HiiragiNetworkWrapper: SimpleNetworkWrapper = NetworkRegistry.INSTANCE.newSimpleChannel(RMReference.MOD_ID)

object HiiragiNetworkManager {

    fun register() {
        HiiragiNetworkWrapper.registerMessage(
            { message: HiiragiMessage.Client, _: MessageContext ->
                syncTile(Minecraft.getMinecraft().world, message)
            },
            HiiragiMessage.Client::class.java,
            0,
            Side.CLIENT
        )
        HiiragiNetworkWrapper.registerMessage(
            { message: HiiragiMessage.Server, ctx: MessageContext ->
                syncTile(ctx.serverHandler.player.world, message)
            },
            HiiragiMessage.Server::class.java,
            1,
            Side.SERVER
        )
        HiiragiNetworkWrapper.registerMessage(
            { message: HiiragiMessage.Primitive, ctx: MessageContext ->
                syncTile(
                    ctx.serverHandler.player.world,
                    message
                ) { tile: TileEntity ->
                    (tile as? TileEntityModuleMachine)?.processRecipe()
                    ctx.serverHandler.player.foodStats.foodLevel -= 1
                }
            },
            HiiragiMessage.Primitive::class.java,
            2,
            Side.SERVER
        )
        HiiragiNetworkWrapper.registerMessage(
            { message: HiiragiMessage.SyncCount, ctx: MessageContext ->
                syncTile(
                    Minecraft.getMinecraft().world,
                    message
                ) { tile: TileEntity -> (tile as? TileEntityModuleMachine)?.currentCount = message.count }
            },
            HiiragiMessage.SyncCount::class.java,
            3,
            Side.CLIENT
        )
        HiiragiNetworkWrapper.registerMessage(
            { message: HiiragiMessage.MinecartTank, _: MessageContext ->
                syncEntity<EntityMinecartTank>(Minecraft.getMinecraft().world, message) { minecart ->
                    minecart.tank.fluid = FluidStack.loadFluidStackFromNBT(message.tag)
                }
                return@registerMessage null
            },
            HiiragiMessage.MinecartTank::class.java,
            4,
            Side.CLIENT
        )
    }

    @Suppress("UNCHECKED_CAST")
    private fun <T : Entity> syncEntity(
        world: World,
        message: IHiiragiMessage.Entity,
        handler: (T) -> Unit
    ): IMessage? {
        world.getEntityByID(message.entityId)?.let { it as? T }?.let(handler)
        return null
    }

    private fun syncTile(
        world: World,
        message: IHiiragiMessage,
        handler: (TileEntity) -> Unit = { tile -> tile.readFromNBT(message.tag) }
    ): IMessage? {
        getTile<TileEntity>(world, message.pos)?.let(handler)
        return null
    }

}